src/arch/riscv/isa.cc

```diff
@@ -984,6 +1003,250 @@ ISA::resetThread()
+        }
+
+        case CSR_HVIP:
+        {
+            INTERRUPT mip = readMiscReg(MISCREG_IP);
+            readval |= (mip.vssi << 2);
+            break;
+        }
+        default: break;
+    }
+
+    return readval;
+}
+
+void
+ISA::writeCSR(uint64_t csr, RegVal writeData)
+{
+    auto csr_it = CSRData.find(csr);
+
+    // panic if the method was used with bad csr idx
+    panic_if(csr_it == CSRData.end(), "Illegal CSR passed to writeCSR");
+
+    // find physical register id (MISCREG == physical)
+    auto midx = csr_it->second.physIndex;
+
+    switch (csr) {
+        case CSR_SIP: case CSR_SIE:
+        case CSR_HIP: case CSR_HIE:
+        {
+            RegVal mideleg = readMiscReg(MISCREG_MIDELEG);
+            writeData &= mideleg;
+            break;
+        }
+
+        // Special case for VSIP & VSIE
+        // We presented the bits in SIP and SIE (one position right)
+        // But we shift them up one again so that they don't overwrite
+        // the actual SIP and SIE but instead take their actual places.
+        // This happens because XIP and XIE are shared registers
+        // but VS-level should be unaware that its SIP and SIE bits are stored
+        // elsewhere.
+        case CSR_VSIP: case CSR_VSIE:
+        {
+            RegVal hideleg = readMiscReg(MISCREG_HIDELEG);
+            writeData <<= 1;
+            writeData &= hideleg;
+            break;
+        }
+        default:
+            break;
+    }
+
+    auto& csr_read_masks = getCSRMaskMap();
+    auto& csr_write_masks = getCSRWriteMaskMap();
+
+    // If no read mask or write mask, assume all bits are writable
+    RegVal write_mask = mask(64);
+
+    // To save space, we only specify write masks that differ
+    // from the respective read masks in misc.hh
+    // If there is no write mask, use the read mask to write
+    // all "visible" aka readable bits. Else use mask(64).
+    auto write_mask_it = csr_write_masks.find(csr);
+    if (write_mask_it != csr_write_masks.end()) {
+        write_mask = write_mask_it->second;
+    } else {
+        auto read_mask_it = csr_read_masks.find(csr);
+        if (read_mask_it != csr_read_masks.end()) {
+            write_mask = read_mask_it->second;
+        }
+    }
+
+    auto writeDataMasked = writeData & write_mask;
+
+    // CSRs are often aliases with different visibility
+    // on the same physical register (MISCREG).
+    // We must keep the values of the non-visible bits
+    // intact for other privileges/contexts where
+    // they are used.
+
+    // Read all the CSR bits
+    auto reg_data_all = backdoorReadCSRAllBits(csr);
+
+    // Only modify those in writeMask
+    auto new_reg_data_all = (reg_data_all & ~write_mask)
+                          | (writeData & write_mask);
+
+
+    switch (csr) {
+        case CSR_FCSR: {
+            setMiscReg(MISCREG_FFLAGS, bits(writeDataMasked, 4, 0));
+            setMiscReg(MISCREG_FRM, bits(writeDataMasked, 7, 5));
+            break;
+        }
```

Comment:
I hope we can align the implement of `MISCREG_VCSR`. They are do similar things.

https://github.com/gem5/gem5/blob/d2c8754ab3b72e52a07d62662de0120b941f1ae5/src/arch/riscv/isa.cc#L595-L599